lib/ansible/plugins/callback/foreman.py

```diff
@@ -80,154 +80,173 @@
         importer. THe data is in a format that Foreman can handle
         without writing another report importer.
         """
-        status = defaultdict(lambda: 0)
-        metrics = {}
-
         for host in stats.processed.keys():
-            sum = stats.summarize(host)
-            status["applied"] = sum['changed']
-            status["failed"] = sum['failures'] + sum['unreachable']
-            status["skipped"] = sum['skipped']
-            log = self._build_log(self.items[host])
-            metrics["time"] = {"total": int(time.time()) - self.start_time}
-            now = datetime.now().strftime(self.TIME_FORMAT)
+            total = stats.summarize(host)
             report = {
-                "report": {
+                "config_report": {
                     "host": host,
-                    "reported_at": now,
-                    "metrics": metrics,
-                    "status": status,
-                    "logs": log,
+                    "reported_at": get_now(),
+                    "metrics": {
+                        "time": {
+                            "total": int(get_time() - self.start_time),
+                        }
+                    },
+                    "status": {
+                        "applied": total['changed'],
+                        "failed": total['failures'] + total['unreachable'],
+                        "skipped": total['skipped'],
+                    },
+                    "logs": list(build_log(self.items[host])),
                 }
             }
-            # To be changed to /api/v2/config_reports in 1.11.  Maybe we
-            # could make a GET request to get the Foreman version & do
-            # this automatically.
+
+            url = self.foreman_url + '/api/v2/config_reports'
             try:
-                r = requests.post(url=self.FOREMAN_URL + '/api/v2/reports',
-                                  data=json.dumps(report),
-                                  headers=self.FOREMAN_HEADERS,
-                                  cert=self.FOREMAN_SSL_CERT,
-                                  verify=self.ssl_verify)
-                r.raise_for_status()
+                response = self.session.post(url=url, json=report)
+                response.raise_for_status()
             except requests.exceptions.RequestException as err:
-                print(to_text(err))
+                self._display.warning(u'Sending report to Foreman at {url} failed for {host}: {err}'.format(
+                    host=host, err=to_text(err), url=self.foreman_url))
+
             self.items[host] = []
 
     def append_result(self, result):
         name = result._task.get_name()
         host = result._host.get_name()
-        self.items[host].append((name, result._result))
+        value = result._result
+        self.items[host].append((name, value))
+        if 'ansible_facts' in value:
+            self.facts[host].update(value['ansible_facts'])
```

Comment:
Maybe?
```suggestion
        self.facts[host].update(value.get('ansible_facts', {}))
```

Reply:
I don't like that in this case. It's a defaultdict which means that if no facts are collected in this run at all, the host will still end up in `self.facts`. The result is that it will `POST` an empty hash to Foreman which removes all facts for that host. This version avoids that by first checking it.